### ISSUE TITLE
Too much catalog columns?

### ISSUE THREAD

**Issue by gforcada** [plone/Products.CMFPlone#1279]
I was about to ask why do we have a ``Type`` and ``portal_type`` catalog columns, but then I took a good look at the [current columns](https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/profiles/default/catalog.xml#L109-L137) and found:

- ``Type`` and ``portal_type`` **and** ``meta_type``
- ``id`` and ``getId``
- ``Creator`` and ``listCreators`` *(ok not a really 1:1 duplication but close enough to mention here)*
- ``getObjSize`` and ``getRemoteUrl`` *(I never used any of them...)*
- ``location`` *(could that be mixed with ``getRemoteUrl``? if actually any of them make sense)*
- ``getIcon`` *(can not that be inferred from Type|portal_type|meta_type?)*

And now for dates... oh man...

First **triplicating, twice,** the same date:
- ``EffectiveDate`` and ``effective`` and ``start``
- ``ExpirationDate`` and ``expires`` and ``end``

And *only* **duplicating, twice again,** some other dates:
- ``ModificationDate`` and ``modified``
- ``CreationDate`` and ``created``

And as with any good joke, the best comes last: after seeing 10 different date names there is yet *another* column:
- ``date`` *(pure gold, who knows if it's about creation, modification effective or expiration, or something completely different...)*

So, in short, I guess a PLIP on reducing the amount of catalog metadata columns used by default would be a good thing, *right?* (note that this last question is rhetorical :-)

I'm not looking at the indexes...

I will create a pull request leaving just one option for each group and see how much breakage we have :D

**Comment by gforcada** [comment]
@plone/framework-team I would like some input from any of you, thanks in advance :-)

**Comment by jensens** [comment]
I'am +1 consolidating and reducing the metadata that was added over time. 

- we should offer some BBB profile for those using addons relying on this metadata
- this is for 5.1, not 5.0.1
- i'd like to see this as a short plip.

my comments:
- start/end is event-related only, while effective/expires is publication releated
- getIcon is in 5.0 a boolean and if theres an icon image or not and can not be inferred from portal_type, but probably from interfaces provided.
- portal_type and meta_type is a different thing, but i dont think we need meta_type any longer. Type is the title of a portal_type iirc and also not needed.
- getRemoteUrl is the url from a Link type and iirc used for immediate redirection in navigation tree.
- getObjSize is probably used in listings and i think its handy to have it.
- location is the DC Location field. (like a city or address). I think we dont need it as metadata.
- date - i have also no idea which date this.

**Comment by thet** [comment]
+1 @gforcada and i agree with all points of @jensens 

When cleaning up the metadata, we should also clean up the catalog indexes.
- ``id`` vs ``getId``
- ``portal_type`` vs ``Type`` vs ``meta_type`` for same reasons mentioned above.
- ``Title`` vs ``sortable_title``
- ``is_folderish`` and ``is_default_page`` could be substituted by interfaces indexed with ``object_provides``
...

**Comment by jensens** [comment]
Title vs sortable_title - well, that makes sense unless we have a new general sortable field index.

**Comment by thet** [comment]
Aha - ``Title`` is a ``ZCTextIndex``, which splits up on each word and creates a tuple of it and thus cannot be sorted, while ``sortable_index`` can, because it's a FieldIndex, which doesn't split up. O.K.

**Comment by hvelarde** [comment]
+1 but lets remember to document it as we can break somebody else's add-ons

**Comment by gforcada** [comment]
thanks for all the comments, I already put aside the ``start`` and ``end`` and updated the pull request accordingly. Even with those we *only* had 94 errors and 1 robot framework test error, so it doesn't look that far from possible.

**Comment by mauritsvanrees** [comment]
Note that there are differences in what the various date metadata do. For example, when `ExpirationDate` is None, `expires` is `2499/12/31 00:00:00 GMT+1`.

I can imagine that the general `Date` column is for the "main" date of an item. So for an event the start date, for Documents it looks like the modification date. I think I may have used this in custom content types. It avoids creating yet another `my_super_special_date` column or index.

Maybe in some of these cases it helps to override the brains with a different class. For example a `PloneBrain` class may have an attribute `getId` that always returns `self.id` and then we can remove the `getId` column.

The `getIcon` metadata has changed recently, in the upcoming Plone 5.0.1, to show thumbnails, right?

**Comment by jensens** [comment]
- I like the idea of a PloneBrain, its a good way to mark things with deprecation warnings.
- I never used metadata ``Date`` in the last 12 (or so) years I'am using Plone, nor did I stumble over code using it.
- getIcon in 5.0 (and consequently implemented now in upcoming 5.0.1) just check if there is an ``image`` attribute on the content and if its there it True, otherwise False.

**Comment by thet** [comment]
I don't see a reason for a Date column, but I like the idea of having a Date index. This could index the creation date of contents, but the start date for events. This can then be used for sorting.
But still, I'd implement something like tihis on a per project basis, because this will fit only a small set of use cases.
+1 for removing, if it is not necessary in core.

@gforcada why are you targeting metadata columns only in your pull request #1280 ? Is it because you want to clean up the catalog in two steps, indexes after columns?

**Comment by gforcada** [comment]
@thet yes, that was by  no means a serious attempt to fix the situation but mostly an exploratory pull request.

My idea was to foster this discussion and have jenkins test what happens if we remove those columns and run tests.

**Comment by idgserpro** [comment]
> Maybe in some of these cases it helps to override the brains with a different class. For example a PloneBrain class may have an attribute getId that always returns self.id and then we can remove the getId column.

+1. We're +1 for consolidating and reducing the metadata that was added over time as well, but at the same time, we're afraid of the *breakage* it may cause and how the BBB code (and for how long) will work. Remember that the same doubt/confusion we're having now "which metadata I should really use", people had it in the past but decided to choose one of them (even the marvelous ```Date```) in their implementation and now we have thousands LOCs using these *obscure* columns. This idea or something in the lines of ```path: (getPhysicalPath accessor method)``` along with the BBB profile that will make getId, getRemoteUrl and such to still work (when possible) is a really nice thing to have.

And about "Type" and "portal_type": we have an explanation of the index in [Plone docs](http://docs.plone.org/develop/plone/searching_and_indexing/indexing.html#default-plone-indexes-and-metadata-columns), but I agree "Type" is too generic, if we're removing confusing metadata and adding a good BBB handling, wouldn't it be the case to rename ambiguous names as well?

For documentation: *we don't know the code and how hard it would be* but I think the explanation for the index/metadata that is now in Plone docs should be at the ZMI itself. Something like:

```xml
 <index name="start" meta_type="DateIndex">
  <property name="index_naive_time_as_local">True</property>
  <property name="description">An optional description here.</property>
 </index>
 <column value="CreationDate">
  <property name="description">An optional description here.</property>
 </column>
```

**Comment by iham** [comment]
as i am currently struggling with indexes ...

besides the names and no more needed indexes/columns:
why is "exclude_from_nav" not a index but a colum?
wouldn't this reduce overhead in global navigation, the navtree portlet, site map, ...?

**Comment by mauritsvanrees** [comment]
I think the column is needed because a Folder that is excluded from navigation may still need to be shown in cases where it has children that are not excluded. So in some cases an index would be helpful, but not always.

**Comment by jensens** [comment]
Still relevant for Plone 6. Only meta_type was removed since 2015. Duplicate columns (CreationDate/created, EffectiveDate/effective, etc.) persist in catalog.xml. Needs a PLIP champion.
